dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityAttributeStore.java

```diff
@@ -176,6 +183,57 @@ public HibernateTrackedEntityAttributeStore( SessionFactory sessionFactory, Jdbc
             .collect( Collectors.toSet() );
     }
 
+    @Override
+    public Set<TrackedEntityAttribute> getAllSearchableAndUniqueTrackedEntityAttributes()
+    {
+        Set<TrackedEntityAttribute> result = new HashSet<>();
+
+        Query programTeaQuery = sessionFactory.getCurrentSession()
+            .createQuery( "select attribute from ProgramTrackedEntityAttribute ptea where ptea.searchable=true" );
+        Query tetypeAttributeQuery = sessionFactory.getCurrentSession()
+            .createQuery(
+                "select trackedEntityAttribute from TrackedEntityTypeAttribute teta where teta.searchable=true" );
+        Query uniqueAttributeQuery = sessionFactory.getCurrentSession()
+            .createQuery( "from TrackedEntityAttribute tea where tea.unique=true" );
+
+        List<TrackedEntityAttribute> programSearchableTrackedEntityAttributes = programTeaQuery.list();
+        List<TrackedEntityAttribute> trackedEntityTypeSearchableAttributes = tetypeAttributeQuery.list();
+        List<TrackedEntityAttribute> uniqueAttributes = uniqueAttributeQuery.list();
+
+        result.addAll( programSearchableTrackedEntityAttributes );
+        result.addAll( trackedEntityTypeSearchableAttributes );
+        result.addAll( uniqueAttributes );
+
+        return result;
+    }
+
+    @Override
+    public boolean createTrigramIndexForAttribute( TrackedEntityAttribute trackedEntityAttribute )
+    {
+        String query = trigramIndexQuery( trackedEntityAttribute );
+        jdbcTemplate.execute( query );
+        return true;
+    }
+
+    @Override
+    public boolean runAnalyzeOnTrackerTables()
+    {
+        jdbcTemplate.execute( ANALYZE_QUERY );
+        return true;
+    }
+
+    @Override
+    public boolean runVacuumOnTrackerTables()
+    {
+        jdbcTemplate.execute( VACUUM_QUERY );
```

Comment:
Do these two (analyze, vacuum) block execution until done?
I wonder what use the boolean return value is. Since it's always true anyway I would return void.

Reply:
No specific use of boolean. Guess it makes sense to return void. 
It does block execution until done. Since its part of a job, I feel that is not a problem?

Reply:
How long can they take? I am not sure if it's a problem as I don't yet know how jobs are implemented in the platform.

Reply:
It does depend on the size of the db. Could be seconds for smaller DB (comparable to our play DBs) and several minutes as the scale of the db increases.